src/mono/mono/metadata/metadata.c

```diff
@@ -7205,27 +7205,35 @@ mono_metadata_load_generic_params (MonoImage *image, guint32 token, MonoGenericC
 		else
 			container->owner.klass = (MonoClass*)real_owner;
 	}
+	/* first pass over the gparam table - just count how many params we own */
+	uint32_t type_argc = 0;
+	uint32_t i2 = i;
+	do {
+		type_argc++;
+		if (++i2 > mono_metadata_table_num_rows (image, MONO_TABLE_GENERICPARAM))
+			break;
+		mono_metadata_decode_row (tdef, i2 - 1, cols, MONO_GENERICPARAM_SIZE);
+	} while (cols [MONO_GENERICPARAM_OWNER] == owner);
+	params = (MonoGenericParamFull *)mono_image_alloc0 (image, sizeof (MonoGenericParamFull) * type_argc);
+
+	/* second pass, fill in the gparam data */
+	mono_metadata_decode_row (tdef, i - 1, cols, MONO_GENERICPARAM_SIZE);
 	do {
 		n++;
-		params = (MonoGenericParamFull *)g_realloc (params, sizeof (MonoGenericParamFull) * n);
-		memset (&params [n - 1], 0, sizeof (MonoGenericParamFull));
 		params [n - 1].owner = container;
```

Comment:
The memset might still be needed.

Reply:
Also, if the number of params is known, this could be a for loop.

Reply:
> The memset might still be needed.

I'm doing `mono_image_alloc0` - each param should be zeroed out already